### PR TITLE
Sketch of the solution for implementing the adapter without `.clone()`.

### PR DESCRIPTION
**Pull request by obi1kenobi**
This compiles fine and doesn't clone `Semantic` or its nodes. I haven't tested it locally but I believe it to be identical.

### CODE REVIEW

**Comment by obi1kenobi** [comment]
Thanks to @nvzqz for suggesting doing `impl<'b> Adapter<'b> for &'b LintAdapter<'a>` which was a key insight here.

File: crates/oxc_linter/src/lint_adapter.rs [hunk 1]
```diff
@@ -715,15 +706,15 @@ impl<'a> Adapter<'a> for LintAdapter<'a> {
...
         edge_name: &Arc<str>,
         _parameters: &EdgeParameters,
         _resolve_info: &ResolveEdgeInfo,
-    ) -> ContextOutcomeIterator<'a, Self::Vertex, VertexIterator<'a, Self::Vertex>> {
+    ) -> ContextOutcomeIterator<'b, Self::Vertex, VertexIterator<'b, Self::Vertex>> {
         let res = match (type_name.as_ref(), edge_name.as_ref()) {
-            ("File", "ast_node") => resolve_neighbors_with(contexts, |v| {
-                v.as_file().unwrap().iter.borrow_mut().take().unwrap()
+            ("File", "ast_node") => resolve_neighbors_with(contexts, |_| {
+                Box::new((*self).semantic.nodes().iter().map(|x| Vertex::AstNode(*x)))
```
Comment (obi1kenobi) [inline crates/oxc_linter/src/lint_adapter.rs:717]: The functionality with the `.clone()` we were trying to eliminate with @Boshen and @u9g is now here, and sans `.clone()`.

Comment (nvzqz) [inline crates/oxc_linter/src/lint_adapter.rs:717]: nit: `(*self)` isn't necessary here

Comment (obi1kenobi) [reply crates/oxc_linter/src/lint_adapter.rs:717]: I figured. I leave it to the maintainers to decide which they'd prefer. Having the explicit dereference helped me convince myself this is valid code, not that I doubt borrowck's wisdom :)

**Pull request merged**